apps/arweave/src/ar_data_sync_worker_master.erl

```diff
@@ -167,84 +183,118 @@ dequeue_main_task(State) ->
 %%--------------------------------------------------------------------
 
 %% @doc If a peer has capacity, take the next task from its queue and schedule it.
-process_peer_queue(Peer, State) ->
-	case peer_has_capacity(Peer, State) andalso peer_has_queued_tasks(Peer, State) of
+process_peer_queue(PeerTasks, State) ->
+	case peer_has_capacity(PeerTasks) andalso peer_has_queued_tasks(PeerTasks) of
 		true ->
-			{sync_range, Args, State2} = dequeue_peer_task(Peer, State),
-			State3 = schedule_sync_range(Peer, Args, State2),
-			process_peer_queue(Peer, State3);
+			{PeerTasks2, sync_range, Args} = dequeue_peer_task(PeerTasks),
+			{PeerTasks3, State2} = schedule_sync_range(PeerTasks2, Args, State),
+			process_peer_queue(PeerTasks3, State2);
 		false ->
-			State
+			{PeerTasks, State}
 	end.
 
+%% @doc the maximum number of tasks we can have in process - including stasks queued here as well
+%% as those scheduled on ar_data_sync_workers.
+max_tasks() ->
+	{ok, Config} = application:get_env(arweave, config),
+	Config#config.sync_jobs * 50.
+
+%% @doc The maximum number of tasks we can have queued for a given peer.
```

Comment:
These changes update the algorithm for deciding when to cut a peer's queue of pending requests.